### PR TITLE
DFU improvements

### PR DESCRIPTION
**Pull request by HiFiPhile**
**Describe the PR**
- Add alt settings support in DFU class, in order to flash multiples partitions (FLASH1, FLASH2, EEPROM, etc.)
- Fix inverted ATTR_MANIFESTATION_TOLERANT logic.

### CODE REVIEW

**Comment by HiFiPhile** [comment]
@xmos-jmccarthy could you take a look if it ok ?

Maybe additional buffer check is needed if transfer size is different for each alt settings.

**Comment by xmos-jmccarthy** [comment]
Good catch on the manifest bit logic.

For the alt setting, can you add an explicit initialized value in `void dfu_moded_init(void)` and the reset value in `dfu_moded_reset`.

**Comment by xmos-jmccarthy** [comment]
> Maybe additional buffer check is needed if transfer size is different for each alt settings.

CFG_TUD_DFU_TRANSFER_BUFFER_SIZE will have to be set to the largest buffer size used by an any given storage type.  The recently added checks to prevent out of bounds memory accesses should be sufficient to catch these cases and alert the user if they make a mistake, but if you could add a note of that to the API header that would be great.

**Comment by HiFiPhile** [comment]
> For the alt setting, can you add an explicit initialized value in void dfu_moded_init(void) and the reset value in dfu_moded_reset.

> you could add a note of that to the API header that would be great.

It's done in cf4220a

File: src/class/dfu/dfu_device.c
Comment (hathach) [inline src/class/dfu/dfu_device.c:193]: the logic is not correct, drv_len keep increasing, while we keep decreasing it which make it more like a division

File: src/class/dfu/dfu_device.c
Comment (hathach) [inline src/class/dfu/dfu_device.c:175]: should be `while(drv_len < max_len)`

File: src/class/dfu/dfu_device.c
Comment (hathach) [inline src/class/dfu/dfu_device.c:187]: - each ATL can have its own attributes, we should store it as an array with `CFG_TUD_DFU_ATL_MAX` (default = 2 or 4). 
- also we need to check to make sure the ALT is consecutive to make sure it is not 2 instances of DFU interface rather than 2 ALT of an DFU instance (hope my words make sense).

File: src/class/dfu/dfu_device.h
Comment (hathach) [inline src/class/dfu/dfu_device.h:70]: need to add `alt` to be consistent as well.

**Review by hathach — CHANGES_REQUESTED** [review]
Thank for the PR, using alt interface as partition doesn't seem to be standard for USB DFU but it is popular with dfu-util. Adding Alt to callback is a good option. Though we need to keep an array of attributes and make sure it is multiple ALTs and not multiple instances of DFUs.

**Comment by HiFiPhile** [comment]
>  using alt interface as partition doesn't seem to be standard for USB DFU

In fact it's in the spec:
```
* Alternate settings can be used by an application to access additional memory segments.  In this case, 
it is suggested that each alternate setting employ a string descriptor to indicate the target memory 
segment;  e.g., “EEPROM”.  Details concerning other possible uses of alternate settings are beyond the 
scope of this document.  However, their use is intentionally not restricted because the authors 
anticipate that implementers will devise additional creative uses for alternate settings.   
```
Something useful but out of spec is add `DFU_DETACH` support in DFU mode, It's workaround for WinUSB who can't issue a Bus reset. 
If there is only one partition MCU can reboot once flash is finished, but if there are multiple partitions MCU don't know when to reboot.

> we need to check to make sure the ALT is consecutive to make sure it is not 2 instances of DFU interface

And I think we can limit support to one interface only, multiple DFU interfaces doesn't make sense I think ?

**Comment by hathach** [comment]
> In fact it's in the spec:
> 
> ```
> * Alternate settings can be used by an application to access additional memory segments.  In this case, 
> it is suggested that each alternate setting employ a string descriptor to indicate the target memory 
> segment;  e.g., “EEPROM”.  Details concerning other possible uses of alternate settings are beyond the 
> scope of this document.  However, their use is intentionally not restricted because the authors 
> anticipate that implementers will devise additional creative uses for alternate settings.   
> ```

Ah right, I totally missed that, thanks for the update.

> And I think we can limit support to one interface only, multiple DFU interfaces doesn't make sense I think ?

Yeah, we did limit the dfu to only once interface, since there is not itf in the callback arguments. However, we still need to check the integrity of the Alt interface with consecutive number. 

> When alt intf is switched, do you think we should reset `state` and `status` to default ?

We probably should, I am not entirely sure. Just make sure you have a bit of explanation comment. We could update it later on when needed.

**Comment by HiFiPhile** [comment]
In fact I missed out that only one functional descriptor is used in DFU class. (Thanks to `dfu-util` maintainer)

I think in this case we need a refactor which expose only one time parameters in functional descriptor. Some VA_ARGS trick like `TUD_BTH_DESCRIPTOR` ?

**Comment by hathach** [comment]
> In fact I missed out that only one functional descriptor is used in DFU class. (Thanks to `dfu-util` maintainer)

Indeed, I totally missed this one as well, I just checked the specs, it will be only 1 Functional DFU descriptor at the end after the series of the Alt Interface

```
ALT 0
ATL 1

ATL n
DFU Functional
```

> I think in this case we need a refactor which expose only one time parameters in functional descriptor. Some VA_ARGS trick like `TUD_BTH_DESCRIPTOR` ?

we probably needs something similar, but I hope we could get a much simpler version.

**Review by hathach — COMMENTED** [review]
I have several comment, but also tried push commit for some of them as well.

File: src/class/dfu/dfu_device.c
Comment (hathach) [reply src/class/dfu/dfu_device.c:254]: I would much more prefer to not store these at all, what is the issue with the existing code ?

File: src/class/dfu/dfu_device.h
Comment (hathach) [inline src/class/dfu/dfu_device.h:54]: also change this from `set_timeout_cb()` to generic `get_status_cb()`

**Review by hathach — COMMENTED** [review]
I push more update to rename get_status_cb(). The PR looks good now, though I still have question regarding your introducing of block and length to the state.

```C
        _dfu_state_ctx.block = request->wValue;
        _dfu_state_ctx.length = request->wLength;
```

File: src/class/dfu/dfu_device.c
Comment (HiFiPhile) [reply src/class/dfu/dfu_device.c:254]: I forgot to delete block and length. With one functional descriptor they are not needed.

Comment (hathach) [reply src/class/dfu/dfu_device.c:254]: I still don't quite understand, why wouldn't we just call the dfu_req_dnload_reply() here as existing code does.

Comment (HiFiPhile) [reply src/class/dfu/dfu_device.c:254]: In order to make the timeout set in `tud_dfu_get_status_cb` work it has to be called before the blocking period of flashing. So I have to delay the flashing callback one step later in the state machine.

Comment (hathach) [reply src/class/dfu/dfu_device.c:254]: It is not correct approach, please revert the change. with the download callback, whenever flashing is complete application should call `tud_dfu_dnload_complete()`. tinyusb will response as busy in the mean time.

**Comment by HiFiPhile** [comment]
No it is the correct approach !

> tinyusb will response as busy in the mean time.

As the CPU is blocked by the Flashing process (Flash is even stalled) how does tinyusb respond ?

Take a look at DFU diagram, device should start flashing after the DFU_GETSTATUS response.
image

And my discussion with `dfu-util` maintainer also confirm this:
>The device should allow the data transfer to finish, and only after replying to the DFU_GETSTATUS it should start flashing and it can then be blocking.

https://sourceforge.net/p/dfu-util/tickets/122/

**Comment by hathach** [comment]
@HiFiPhile oh, I see, device need to response to get status first before performing the flashing. In that case, It make sense

**Comment by HiFiPhile** [comment]
> @HiFiPhile oh, I see, device need to response to get status first before performing the flashing. In that case, It make sense

Yes, it's exactly the case.

**Comment by hathach** [comment]
> > @HiFiPhile oh, I see, device need to response to get status first before performing the flashing. In that case, It make sense
> 
> Yes, it's exactly the case.

I am doing more rename and update to DFU, will push when complete.

**Comment by hathach** [comment]
@HiFiPhile it need a bit more work, the get_status should allow application to report error such as incorrect address (such error can be from incorrect download file from host). I will do a bit more update for that

image

**Comment by HiFiPhile** [comment]
> it need a bit more work, the get_status should allow application to report error such as incorrect address (such error can be from incorrect download file from host).

Ok,  errAddress could be useful.

**Comment by hathach** [comment]
@HiFiPhile @xmos-jmccarthy I have updated with lots of rename (minor) and also tweak behavior as well as flow for DFU callback . I also migrated away from state machine since I found it a bit more complicated (left in code for review only). In sum up
- `get_timeout()` will be invoked (by GET_STATUS) before download_cb() and manifest_cb(). Since either has potential flashing op.
- `finish_flashing(status)` ( renamed from download_complete() ) is called from application after download/manifest cb() after flashing is complete. should address out of bound, or flash & verified failed, status can be set to reflect the error. DFU will then report that to host and enter dfuError state.
-  manifest_cb() replace both firmware_valid_check_cb() / device_data_done_check_cb(). It can be used to flash (if buffered whole image so far) or perform checksum and still can report error.

**Comment by HiFiPhile** [comment]
I think it's in good form with a quick look, will do some real world tests.

**Comment by xmos-jmccarthy** [comment]
I like the callback changes, this is much nicer to look at.

I will update my port and run some tests on hardware.  I should be able to do that next week at the latest.

**Comment by hathach** [comment]
I will also integrate this to tinyuf2 project as well to see if there is any issue with existing API https://github.com/adafruit/tinyuf2

**Comment by HiFiPhile** [comment]
I've done some tests on LPC4337, so far so good.

**Comment by hathach** [comment]
@HiFiPhile I only use auto format with portion of text selection where I want to. Normally I will just manual type it. Auto format is not enabled, since I couldn't figure out how to get it auto align `=`. Also it is a bit too strict, as long as the code look good, it is ok to go over line boundary (which I set to 120) or break rule. It is not too important, don't worry too much about it.

**Review by hathach — APPROVED** [review]
Thank you @HiFiPhile for stating this PRs, and @xmos-jmccarthy for helping with the review. This PR is now ready, and I am glad that you are both happy with the changes. There is quite a bit of rename though, but it is to make it more like other class. If there is nothing else that need changes, I wink we could merge this. Any issues can be fixed as follow up PRs.

**Review by xmos-jmccarthy — APPROVED** [review]
Everything looks good.

**Comment by HiFiPhile** [comment]
It's good for me :)

**Pull request merged**
